development-tools/warnings.rst

```diff
@@ -0,0 +1,69 @@
+.. warnings:
+
+Tools for tracking compiler warnings
+====================================
+
+The compiler warning tracking tooling is intended to alert developers to new
+compiler warnings introduced by their contributions. The tooling consists of
+Python script which is ran by a few GitHub Actions workflows. These
+GitHub Actions jobs run the warning checking tooling:
+
+- Ubuntu/Build and Test
+- macOS/Build and Test
+
+The help text for the :cpy-file:`Tools/build/check_warnings.py` is as follows:
+
+.. code-block:: text
+
+    usage: check_warnings.py [-h]
+           -c COMPILER_OUTPUT_FILE_PATH
+           [-i WARNING_IGNORE_FILE_PATH] [-x] [-X] -t {json,clang}
+
+    options:
+        -h, --help            show this help message and exit
+        -c COMPILER_OUTPUT_FILE_PATH, --compiler-output-file-path COMPILER_OUTPUT_FILE_PATH
+                                                    Path to the compiler output file
+        -i WARNING_IGNORE_FILE_PATH, --warning-ignore-file-path WARNING_IGNORE_FILE_PATH
+                                                    Path to the warning ignore file
+        -x, --fail-on-regression
+                                                    Flag to fail if new warnings are found
+        -X, --fail-on-improvement
+                                                    Flag to fail if files that were expected to have warnings have no warnings
+        -t {json,clang}, --compiler-output-type {json,clang}
+                                                    Type of compiler output file (json or clang)
+
+The script can be run locally by providing the compiler output and the
+compiler output type to see a list of unique warnings.
+
+::
+
+        $ python Tools/build/check_warnings.py -c <compiler_output_file> -t <compiler_output_type>
+
+..
+
+    Note: `-fdiagnostics-format=json` flag is required when compiling with GCC
+           for the script to properly parse the compiler output.
+
+.. _warning-check-failure:
+
+What to do if a warning check fails GitHub CI
+---------------------------------------------
+
+The warning check tooling will fail if the compiler generates more or less
+warnings than expected for a given source file as defined in the
+platform-specific warning ignore file. The warning ignore file is either
+:cpy-file:`Tools/build/.warningignore_ubuntu` or :cpy-file:`Tools/build/.warningignore_macos`
+depending on the platform.
```

Comment:
Is `:cpy-fil:`!check_warnings.py` supposed to be shorthand to link to a file with the same basename already referenced in the document? When build the dev guide with the suggestion it would take me to `cpython/check_warnings.py`.

I put in the full path for now.

Reply:
The `!` tells Sphinx not to generate a link, since for brevity I only used the filename and not the full path, a link can't be generated.  By still using `:cpy-file:` the filename is semantically marked and has the same style as the other filenames marked with the same role.  The file is already linked above, so there is no need to repeat the link.

Note that an alternative would have been `` :cpy-file:`check_warnings.py <Tools/build/check_warnings.py>` ``, which would have linked to the correct path while only showing the filename, but that doesn't seem to be supported by the `:cpy-file:` role.

Reply:
```
:cpy-file:`!check_warnings.py`
```

Renders as

<img width="309" alt="Screenshot 2024-08-10 at 7 34 43 PM" src="https://github.com/user-attachments/assets/d7a03d48-1de0-4c98-84a2-13ea8a55ef99">

And actually links to `cpython/check_warnings`. Is that what you are suggesting?